Rpackage/R/fitAndScore.R

```diff
@@ -0,0 +1,204 @@
+    for(i in 1 : n_folds){
+      fold_data <- mmit.pruning(fold_tree[[i]])
+      fold_alphas[[i]] <- rev(unlist(lapply(fold_data, function(x) x$alpha)))
+      fold_prune_trees[[i]] <- rev(lapply(fold_data, function(x) x$tree)) 
+    }
+    
+    # Compute the test risk for all pruned trees of each fold
+    alpha_path_score <- NULL
+    for(i in 1 : n_folds){
+      
+      #dummy value initialised to form dataframe, removed later
+      alpha_path_score[[i]] <- c(0, 0, 0)   
+      
+      for(j in 1 : length(fold_alphas[[i]])){
+        ### convert pruned tree list to partynode
+        node <- fold_prune_trees[[i]][[j]]
+        
+        ### predictions of the tree for test data
+        prediction <- mmit.predict(node, feature.mat[fold_split_idx$test[i,],])
+
+        ###error calc
+        fold_test_scores <- scorer(target.mat[fold_split_idx$test[i,],], prediction)
+        
+        ### creating a dataframe with init alpha, final alpha, score value
+        if(j < length(fold_alphas[[i]])){
+          data <- c(fold_alphas[[i]][j], fold_alphas[[i]][j + 1], fold_test_scores)
+          alpha_path_score[[i]] <- rbind(data, alpha_path_score[[i]])
+        }
+        else{
+          data <- c(fold_alphas[[i]][j], Inf, fold_test_scores)
+          alpha_path_score[[i]] <- rbind(data, alpha_path_score[[i]])
+        }
+      }
+      
+      colnames(alpha_path_score[[i]]) <- c("init alpha", " final alpha", "score")
+      alpha_path_score[[i]] <- as.data.frame(alpha_path_score[[i]], row.names = "")
+      alpha_path_score[[i]] <- head(alpha_path_score[[i]], -1)
+    }
+    
+    # Prune the master tree based on the CV estimates
+    alphas <- NULL
+    alpha_cv_scores <- NULL
+    best_alpha <- -1
+    best_score <- attr(scorer, "worst")
+    best_tree <- NULL
+    
+    for(i in 1 : length(master_alphas)){
+      if(i < length(master_alphas)){
+        geo_mean_alpha_k <- sqrt(master_alphas[[i]] * master_alphas[[i + 1]])
+      }
+      else{
+        geo_mean_alpha_k <- Inf
+      }
+
+      ### calc train score
+      prediction <- mmit.predict(master_pruned_trees[[i]], feature.mat)
+      train_score <- scorer(target.mat, prediction)
+      
+      ### calc cost of all leaves
+      ter_id <- nodeids(master_pruned_trees[[i]], terminal = TRUE)
+      n <- nodeapply(master_pruned_trees[[i]], ids = ter_id, info_node)
+      train_objective <- sum(matrix(unlist(n), nrow = length(n), byrow = T)[, 2])
+      
+      alpha_train_scores <- c(alpha_train_scores, train_score)
+      alpha_train_objective_values <- c(alpha_train_objective_values, train_objective)
+      
+      ### compute cv_score as mean of each fold scores
+      cv_score <- 0
+      for(j in 1 : n_folds){
+        for(k in 1 : nrow(alpha_path_score[[j]])){
+          if((geo_mean_alpha_k < alpha_path_score[[j]][k, ]$` final alpha`) && (geo_mean_alpha_k >= alpha_path_score[[j]][k, ]$`init alpha`)){
+            cv_score <- cv_score + alpha_path_score[[j]][k, ]$score
```

Comment:
In the CV results you have some 0 values in cv_score, which is highly surprising. These 0 values are always associated with alpha = inf. I suspect that this is a bug due to the reverse sorting of the alphas and trees. Should verify if this still occurs after correcting the bug.

Reply:
So as you said, alphas are in correct order, but I still find it odd to have some cv scores equal to 0. I'll debug to see if it happens that you don't get into the loop

Reply:
Can you please verify that the score is indeed updated for each fold? This can be done by setting a variable to False at the beginning of the loop of a fold (e.g., line 122) and setting that variable to True once the condition is satisfied. Then at the end of the fold loop, you can assert that the condition was satisfied at least once for the fold.

I suspect that sometimes the code doesnt go into the condition and thus the score is left at 0.

Reply:
yeah...possibly it doeant go into the loop for alpha = Inf, i'll check

Reply:
found the bug, it was not going inside the loop due to this statement
`if((geo_mean_alpha_k < alpha_path_score[[j]][k, ]$final alpha` 
When geo_mean_alpha_k  = Inf this gives false even for final alpha = Inf, thus it doesnt go inside the loop.

Reply:
But what values does the between dict consist of? is it initial alpha to final alpha excluding final alpha or including, because in next (j+1th) iteration the final alpha of j is initial alpha of j+1th loop

Reply:
So in the Python implementation, the between dict is used to find the pruned tree that corresponds to the range [alpha_k, alpha_(k+1)[ (notice exclusion of upper limit). So if the geomean is in that range, the corresponding tree is returned. You can see that the between dict has special conditions for infinite values: https://github.com/aldro61/mmit/blob/master/mmit/utils.py#L39. Basically, you should reproduce this condition in your code. In the new code (after your bug fix), you have the condition `geo_mean_alpha_k <= alpha_path_score[[j]][k, ]$` final alpha`, which I think is incorrect, because of the `=` and the fact that the upper limit should be excluded.

Reply:
rectified